partition-manager/generation/src/lib.rs

```
@@ -0,0 +1,350 @@
+            let s = String::deserialize(deserializer)?;
+            Ok(if s == "any" { Variant::Any } else { Variant::Other(s) })
+        }
+    }
+
+    impl From<String> for Variant {
+        fn from(value: String) -> Self {
+            Variant::Other(value)
+        }
+    }
+
+    impl From<&str> for Variant {
+        fn from(value: &str) -> Self {
+            Variant::Other(value.to_string())
+        }
+    }
+
+    #[derive(Debug, Clone, Copy, PartialEq, Deserialize)]
+    pub enum Access {
+        #[serde(alias = "ro")]
+        #[serde(alias = "read-only")]
+        RO,
+        #[serde(alias = "rw")]
+        #[serde(alias = "read-write")]
+        RW,
+    }
+
+    #[derive(Debug, Clone, PartialEq, Deserialize)]
+    #[doc(hidden)]
+    pub struct Partition {
+        #[serde(default)]
+        pub access: BTreeMap<Variant, Access>,
+        pub offset: u32,
+        pub size: u32,
+    }
+
+    #[allow(clippy::from_over_into)]
+    impl Into<Range<u32>> for &Partition {
+        fn into(self) -> Range<u32> {
+            self.offset..(self.offset + self.size)
+        }
+    }
+
+    fn ranges_overlap(l: &Range<u32>, r: &Range<u32>) -> bool {
+        l.start < r.end && r.start < l.end
+    }
+
+    impl Partition {
+        pub fn overlaps(&self, other: &Partition) -> bool {
+            ranges_overlap(&self.into(), &other.into())
+        }
+    }
+
+    #[doc(hidden)]
+    #[derive(Debug, PartialEq)]
+    pub struct GeneratedPartition {
+        pub name: String,
+        pub access: Access,
+        pub offset: u32,
+        pub size: u32,
+    }
+
+    impl GeneratedPartition {
+        pub fn name_access(self) -> (String, Access) {
+            (self.name, self.access)
+        }
+    }
+
+    #[derive(Debug, Clone, PartialEq, Deserialize)]
+    #[doc(hidden)]
+    pub struct Manifest {
+        #[serde(default)]
+        pub variants: BTreeSet<Variant>,
+        pub disk: Disk,
+        pub partitions: BTreeMap<String, Partition>,
```

I think this will silently drop `Partition`s when there are redundant partition names.

---

Hmm, I am not 100% convinced this is something we should handle, but I added a custom deserializer to handle this just in case.